Source/Core/AsyncQueue.cs

```diff
@@ -0,0 +1,70 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Boogie;
+
+/// <summary>
+/// A queue from which items can be dequeued even if they have not been enqueued yet.
```

Comment:
I don't understand that description

Reply:
> I wonder if the queue could just be implemented with a semaphore. You would just release it when enqueuing and wait on it when dequeing.

Wouldn't the dequeue become synchronous then?

Reply:
Yeah, I would have just locked a regular queue.  Something like `await semaphore.WaitAsync(); lock(q) { q.Deque(); }`?

Reply:
It works well, and it makes AsyncQueue so simple I might remove it altogether. Thanks.

Reply:
One downside of the `await semaphore.WaitAsync(); lock(q) { q.Dequeue(); }` approach is that the items don't come out in the same order as they're put in. For `CheckerPool` that is fine but for `SMTLibProcess` a little less so.

Reply:
How is that? Isn't dequeuing unloading items the same order?

Reply:
The test `ItemsAreKeptInOrder` fails in the implementation that uses a SemaphoreSlim, in commit https://github.com/boogie-org/boogie/pull/552/commits/5db0f4dcdaab49f9d605dbf980e02deb5677f278

Reply:
Oh I see. It's probably because `await semaphore.WaitAsync(); ...` does not guarantee that the first thread to reach the semaphore will be triggered next? In that case, you might consider to keep your previous code.

Reply:
> Oh I see. It's probably because await semaphore.WaitAsync(); ... does not guarantee that the first thread to reach the semaphore will be triggered next? In that case, you might consider to keep your previous code.

Yes that.